velox/type/DecimalUtil.h

```diff
@@ -32,5 +32,39 @@ class DecimalUtil {
   /// Helper function to convert a decimal value to string.
   template <typename T>
   static std::string toString(const T& value, const TypePtr& type);
+
+  inline static int128_t rescaleWithRoundUp(
+      const int128_t& unscaledValue,
+      const uint8_t fromScale,
+      const std::pair<uint8_t, uint8_t>& toPrecisionScale,
+      bool& isNullOut) {
+    auto rescaledValue = unscaledValue;
+    auto rf = toPrecisionScale.second - fromScale;
+    if (rf >= 0) {
+      rescaledValue = unscaledValue * DecimalUtil::kPowersOfTen[rf];
+    } else {
+      rf = -1 * rf;
+      rescaledValue = unscaledValue / DecimalUtil::kPowersOfTen[rf];
+      int128_t remainder = unscaledValue % DecimalUtil::kPowersOfTen[rf];
+      if (unscaledValue >= 0) {
+        if (remainder > DecimalUtil::kPowersOfTen[rf] / 2) {
+          rescaledValue++;
+        }
+      } else {
+        if (remainder < -DecimalUtil::kPowersOfTen[rf] / 2) {
+          rescaledValue--;
+        }
+      }
+    }
+    // Check overflowing
+    auto valueToCompare = rescaledValue;
+    if (rescaledValue < 0) {
+      valueToCompare *= -1;
+    }
+    if (valueToCompare > DecimalUtil::kPowersOfTen[toPrecisionScale.first]) {
```

Comment:
This is an invalid state than can be caused due to an invalid input correct? Then we should VELOX_DCHECK here and not quietly return null.

Reply:
I see the test where a valid input might not fit in the expected type. So, this check makes sense.

Reply:
Yes, I tried to keep the behavior similar to Presto. Presto does this as well.

Reply:
I see Presto throws an error. Can you verify?
```
 Cannot cast DECIMAL '123.45000' to DECIMAL(3, 1)
```

Reply:
I think the scale can be decreased. But not precision in the `toType`